README.md
# express-chaos-middleware

Put some chaos in your express application.

## Description

The aim of this library is to provide simple chaos testing cases for express applications.

## Installation

```sh
npm add express-chaos-middleware
```

## Features

- Requests slowdown : Endpoints can be slowed down randomly 
- Random response error : Endpoints can respond randomly with an error code
- Random exception : Endpoints can throw exceptions randomly
- Seeding : Manage random to replay scenarios

Chaos features have a 10 percent "chance" of being called by default. The percentage can be changed in the configuration object.

## Seeding ?

A seed is a value (alphanumeric in this module) used to created a pseudorandom number generator.

The seed is used as the starting point for generating a sequence of random numbers that appear to be random, but are actually deterministic based on the seed value.

In our case, this is really useful to replay scenarios.

Example :
- The first API call trigger the slow function during 250ms
- The second call is normal (classic result)
- The third throw an exception

With a classic random function, it's impossible to replay this use case.

On the other hand, it's possible to replay this use case with a seeding system by using the same seed.

See https://en.wikipedia.org/wiki/Random_seed.

## Usage

### Simple usage

```js
const {chaos} = require('express-chaos-middleware');

app.use(chaos())
```

### Use an existing seed

The random seed can be configured in the input object param.

```js
const {chaos} = require('express-chaos-middleware');

app.use(chaos({
  // this seed throw an exception on the first call
  seed: 'uJaK8BUr2084pph',
}))
```

### Advanced usage - Change chaos probability

The probability of chaos can be configured in the input object param.

This parameter is a number value between 0 and 100.

```js
const {chaos} = require('express-chaos-middleware');

app.use(chaos({
  probability: 50,
}))
```

### Advanced usage - Override max delay

The max slowdown delay can be configured in the input object param.

```js
const {chaos} = require('express-chaos-middleware');

app.use(chaos({
  maxDelay: 1000,
}))
```

### Advanced usage - Use custom http error codes array

The http error codes array can be overridden in the input object param.

```js
const {chaos} = require('express-chaos-middleware');

app.use(chaos({
  errCodes: [400, 500],
}))
```

### Advanced usage - Use custom feature list 

The chaos feature list can be overridden in the input object param.

You can for example only use delay and http error codes with the following configuration

All rules are exported in the `Rules` object.

```js
const {chaos} = require('express-chaos-middleware');

app.use(chaos({
  rules: [Rules.DELAY, Rules.HTTPERROR],
}))
```



CHANGELOG.md
## [1.2.4](https://github.com/drouian-m/express-chaos-middleware/compare/v1.2.3...v1.2.4) (2023-04-05)


### Bug Fixes

* **seed:** use debug instead of console log ([eb6edc6](https://github.com/drouian-m/express-chaos-middleware/commit/eb6edc64a76bcbea0e5e9ef788156d5b3aad70cf))

## [1.2.3](https://github.com/drouian-m/express-chaos-middleware/compare/v1.2.2...v1.2.3) (2023-04-05)


### Bug Fixes

* **npm:** update ignore file ([d2c8e31](https://github.com/drouian-m/express-chaos-middleware/commit/d2c8e314c91694ec58ba318bc87f9f929f904e05))

## [1.2.2](https://github.com/drouian-m/express-chaos-middleware/compare/v1.2.1...v1.2.2) (2023-04-05)


### Bug Fixes

* **ci:** build ts before publish ([752357c](https://github.com/drouian-m/express-chaos-middleware/commit/752357c82ff70e26d19dc5581996a10ed46fc101))

## [1.2.1](https://github.com/drouian-m/express-chaos-middleware/compare/v1.2.0...v1.2.1) (2023-04-05)


### Bug Fixes

* **npm:** update main file ([92c22dd](https://github.com/drouian-m/express-chaos-middleware/commit/92c22dd226652a473a6b562d7a621be2db1108aa))

# [1.2.0](https://github.com/drouian-m/express-chaos-middleware/compare/v1.1.4...v1.2.0) (2023-04-05)


### Features

* **seed:** add random seeds ([ad3e754](https://github.com/drouian-m/express-chaos-middleware/commit/ad3e7542f64b881e0b8a1162dbdbc75d12c88534))

## [1.1.4](https://github.com/drouian-m/express-chaos-middleware/compare/v1.1.3...v1.1.4) (2022-12-22)


### Bug Fixes

* **ci:** restore changelog ([11dd1f2](https://github.com/drouian-m/express-chaos-middleware/commit/11dd1f22780a2c9ae6a61197c1be03cf8d14d998))

# [1.1.0](https://github.com/drouian-m/express-chaos-middleware/compare/v1.0.0...v1.1.0) (2022-12-21)


### Features

* **wait:** replace timer promise ([4c7d30f](https://github.com/drouian-m/express-chaos-middleware/commit/4c7d30fcce67c7b841066800f04398d706d0538e))

# 1.0.0 (2021-12-20)


### Features

* **ci:** add release workflow ([f58ca75](https://github.com/drouian-m/express-chaos-middleware/commit/f58ca75d05a1c333a70075a64a08d3f4fb0080c2))
* init project ([b4390fa](https://github.com/drouian-m/express-chaos-middleware/commit/b4390fa24964e6ebd25e3f98d59872270467ca0e))


src/index.spec.js
const lib = require('./index');

describe('Express chaos middleware', () => {
  beforeEach(() => jest.clearAllMocks);
  it('should slow application', (done) => {
    const randomSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.1);
    const randomRulesSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0);
    const randomDelaySpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.1);
    const delaySpy = jest.spyOn(lib.Rules, 'DELAY');
    const req = jest.fn();
    const res = jest.fn();
    const next = () => {
      expect(randomSpy).toHaveBeenCalled();
      expect(randomRulesSpy).toHaveBeenCalled();
      expect(randomDelaySpy).toHaveBeenCalled();
      expect(delaySpy).toHaveBeenCalled();
      done();
    };

    lib.chaos()(req, res, next);
  });

  it('should return http error', (done) => {
    const randomSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.1);
    const randomRulesSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.4);
    const httpErrSpy = jest.spyOn(lib.Rules, 'HTTPERROR');
    const req = jest.fn();
    const res = {
      statusCode: 200,
      end: () => {
        expect(randomSpy).toHaveBeenCalled();
        expect(randomRulesSpy).toHaveBeenCalled();
        expect(httpErrSpy).toHaveBeenCalled();
        done();
      },
    };

    lib.chaos()(req, res);
  });

  it('should throw error', () => {
    const randomSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.1);
    const randomRulesSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.8);
    const exceptionSpy = jest.spyOn(lib.Rules, 'EXCEPTION');
    const req = jest.fn();
    expect(() => lib.chaos()(req)).toThrow(Error);
    expect(randomSpy).toHaveBeenCalled();
    expect(randomRulesSpy).toHaveBeenCalled();
    expect(exceptionSpy).toHaveBeenCalled();
  });

  it('should skip chaos function', (done) => {
    const randomSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.5);
    const req = jest.fn();
    const res = jest.fn();
    const next = () => {
      expect(randomSpy).toHaveBeenCalled();
      done();
    };

    lib.chaos()(req, res, next);
  });

  it('should override probability', (done) => {
    const randomSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.8);
    const req = jest.fn();
    const res = jest.fn();
    const next = () => {
      expect(randomSpy).toHaveBeenCalled();
      done();
    };

    lib.chaos({
      probability: 75,
    })(req, res, next);
  });

  it('should override rules', (done) => {
    const randomSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.1);
    const randomRulesSpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0);
    const randomDelaySpy = jest.spyOn(Math, 'random').mockImplementationOnce(() => 0.1);
    const delaySpy = jest.spyOn(lib.Rules, 'DELAY');
    const req = jest.fn();
    const res = jest.fn();
    const next = () => {
      expect(randomSpy).toHaveBeenCalled();
      expect(randomRulesSpy).toHaveBeenCalled();
      expect(randomDelaySpy).toHaveBeenCalled();
      expect(delaySpy).toHaveBeenCalled();
      done();
    };

    lib.chaos({
      rules: [lib.Rules.DELAY],
      maxDelay: 200,
    })(req, res, next);
  });

  it('should throw error if probability value is invalid (not a number)', () => {
    expect(() => lib.chaos({ probability: 'a' })).toThrow(Error);
  });

  it('should throw error if probability value is invalid ( < 0 )', () => {
    expect(() => lib.chaos({ probability: -5 })).toThrow(Error);
  });

  it('should throw error if probability value is invalid ( > 100 )', () => {
    expect(() => lib.chaos({ probability: 200 })).toThrow(Error);
  });
});


src/index.spec.ts
import { describe, expect, beforeEach, it, vi } from "vitest";
import type { Request, Response, NextFunction } from "express";
import * as lib from "./index";

describe("Express chaos middleware", () => {
	beforeEach(() => vi.clearAllMocks);
	it("should slow application", () =>
		new Promise<void>((done) => {
			const delaySpy = vi.spyOn(lib.Rules, "DELAY");
			const req = vi.fn() as unknown as Request;
			const res = vi.fn() as unknown as Response;
			const next = () => {
				expect(delaySpy).toHaveBeenCalled();
				done();
			};

			lib.chaos({ seed: "nKHGgrOtnZzTrXk" })(req, res, next);
		}));

	it("should return http error", () =>
		new Promise<void>((done) => {
			const httpErrSpy = vi.spyOn(lib.Rules, "HTTPERROR");
			const req = vi.fn() as unknown as Request;
			const res = {
				statusCode: 200,
				status: (code) => {
					//@ts-ignore
					this.statusCode = code;
				},
				end: () => {
					expect(httpErrSpy).toHaveBeenCalled();
					done();
				},
			} as Response;
			const next = () => true;

			lib.chaos({
				seed: "ApOlNU3K89qZ8AJ",
			})(req, res, next);
		}));

	it("should throw error", () => {
		const exceptionSpy = vi.spyOn(lib.Rules, "EXCEPTION");
		const req = vi.fn() as unknown as Request;
		const res = vi.fn() as unknown as Response;
		const next = vi.fn() as unknown as NextFunction;

		expect(() =>
			lib.chaos({
				seed: "uJaK8BUr2084pph",
			})(req, res, next),
		).toThrow(Error);
		expect(exceptionSpy).toHaveBeenCalled();
	});

	it("should skip chaos function", () =>
		new Promise<void>((done) => {
			const exceptionSpy = vi.spyOn(lib.Rules, "EXCEPTION");
			const httpErrSpy = vi.spyOn(lib.Rules, "HTTPERROR");
			const delaySpy = vi.spyOn(lib.Rules, "DELAY");
			const req = vi.fn() as unknown as Request;
			const res = vi.fn() as unknown as Response;
			const next = () => {
				expect(exceptionSpy).not.toHaveBeenCalled();
				expect(httpErrSpy).not.toHaveBeenCalled();
				expect(delaySpy).not.toHaveBeenCalled();
				done();
			};

			lib.chaos({ seed: "LcjNZUJiBYyXmpQ" })(req, res, next);
		}));

	it("should override probability", () =>
		new Promise<void>((done) => {
			const delaySpy = vi.spyOn(lib.Rules, "DELAY");
			const req = vi.fn() as unknown as Request;
			const res = vi.fn() as unknown as Response;
			const next = () => {
				expect(delaySpy).toHaveBeenCalled();
				//@ts-ignore
				done();
			};

			lib.chaos({
				seed: "Bn4tHP7XNDLYtsk",
				probability: 75,
			})(req, res, next);
		}));

	it("should override rules", () =>
		new Promise<void>((done) => {
			const delaySpy = vi.spyOn(lib.Rules, "DELAY");
			const req = vi.fn() as unknown as Request;
			const res = vi.fn() as unknown as Response;
			const next = () => {
				expect(delaySpy).toHaveBeenCalled();
				//@ts-ignore
				done();
			};

			lib.chaos({
				seed: "HXoXdi8CPxHiLLd",
				rules: [lib.Rules.DELAY],
				maxDelay: 200,
			})(req, res, next);
		}));

	it("should throw error if probability value is invalid (not a number)", () => {
		//@ts-ignore
		expect(() => lib.chaos({ probability: "a" })).toThrow(Error);
	});

	it("should throw error if probability value is invalid ( < 0 )", () => {
		expect(() => lib.chaos({ probability: -5 })).toThrow(Error);
	});

	it("should throw error if probability value is invalid ( > 100 )", () => {
		expect(() => lib.chaos({ probability: 200 })).toThrow(Error);
	});
});


src/index.js
const debug = require('debug')('express-chaos-middleware');

/**
 * @typedef {object} ChaosOptions
 * @property {number} [probability] Issue probability percent (0-100)
 * @property {number} [maxDelay] Max delay (ms)
 * @property {Array<number>} [errCodes] Error codes array
 * @property {Array<Rules>} [rules] Rules to apply
 */

/**
 * Wait X ms
 */
const wait = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Requests slowdown
 */
async function delay(req, res, next) {
  const maxDelay = req.options?.maxDelay || 500;
  const pause = Math.floor(Math.random() * maxDelay);
  debug(`rulesFunctions.delay - wait ${pause} ms`);
  await wait(pause);
  next();
}

/**
 * Random response error
 */
function httpError(req, res) {
  const errCodes = req.options?.errCodes || [400, 401, 403, 404, 409, 500];
  const code = errCodes[Math.floor(Math.random() * errCodes.length)];
  debug(`rulesFunctions.httpError - send ${code} code`);
  res.statusCode = code;
  res.end();
}

/**
 * Throw random exception
 */
function exception() {
  debug(`rulesFunctions.exception - throw exception`);
  throw new Error('BOOM !');
}

const Rules = {
  DELAY: delay,
  HTTPERROR: httpError,
  EXCEPTION: exception,
};

/**
 * Randomly throw error and slow response
 * @param {ChaosOptions} options
 * @returns {Promise<void>}
 */
function chaos(options) {
  const probability = options?.probability || 10;
  if (!Number.isInteger(probability) || probability < 0 || probability > 100) {
    throw new Error('Invalid probability value');
  }

  const rules = options?.rules || [Rules.DELAY, Rules.HTTPERROR, Rules.EXCEPTION];
  return (req, res, next) => {
    req.options = options;
    const rand = Math.floor(Math.random() * 100);
    if (rand <= probability) {
      const rule = rules[Math.floor(Math.random() * rules.length)];
      return rule(req, res, next);
    }

    return next();
  };
}

module.exports = { chaos, Rules };


src/index.ts
import Debug from "debug";
import seedrandom from "seedrandom";

import type { Request, Response, NextFunction } from "express";
const debug = Debug("express-chaos-middleware");

const Rules = {
	DELAY: delay,
	HTTPERROR: httpError,
	EXCEPTION: exception,
};

type UserOptions = {
	seed?: string;
	probability?: number;
	maxDelay?: number;
	errCodes?: number[];
	rules?: Function[];
};

type ChaosOptions = UserOptions & {
	rng: Function;
};

/**
 * Generate alphanumeric seed
 * @param length - seed length
 * @returns seed
 */
function generateSeed(length: number): string {
	var result = "";
	var characters =
		"ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
	var charactersLength = characters.length;
	for (var i = 0; i < length; i++) {
		result += characters.charAt(Math.floor(Math.random() * charactersLength));
	}
	return result;
}

/**
 * Wait X ms
 * @param ms - waiting time
 */
const wait = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/**
 * Requests slowdown
 * @param req - express request
 * @param res - express response
 * @param next - express next function
 */
async function delay(req: Request, res: Response, next: NextFunction) {
	const maxDelay = req.options?.maxDelay || 500;
	const pause = Math.floor(req.options.rng() * maxDelay);
	debug(`rulesFunctions.delay - wait ${pause} ms`);
	await wait(pause);
	next();
}

/**
 * Random response error
 * @param req - express request
 * @param res - express response
 */
function httpError(req: Request, res: Response) {
	const errCodes = req.options?.errCodes || [400, 401, 403, 404, 409, 500];
	const code = errCodes[Math.floor(req.options.rng() * errCodes.length)];
	debug(`rulesFunctions.httpError - send ${code} code`);
	res.status(code);
	res.end();
}

/**
 * Throw random exception
 */
function exception() {
	debug(`rulesFunctions.exception - throw exception`);
	throw new Error("BOOM !");
}

/**
 * Randomly throw error and slow response
 * @param options - user options
 * @returns a chaos middleware instance
 */
function chaos(options?: UserOptions) {
	const seed = options?.seed || generateSeed(15);
	debug(`current seed ${seed}`);
	const rng = seedrandom(seed);
	const chaosOptions: ChaosOptions = {
		...options,
		seed,
		rng,
	};

	const probability = chaosOptions?.probability || 10;
	if (!Number.isInteger(probability) || probability < 0 || probability > 100) {
		throw new Error("Invalid probability value");
	}

	const rules = chaosOptions?.rules || [
		Rules.DELAY,
		Rules.HTTPERROR,
		Rules.EXCEPTION,
	];

	return (req: Request, res: Response, next: NextFunction) => {
		req.options = chaosOptions;
		const rand = Math.floor(chaosOptions.rng() * 100);
		if (rand <= probability) {
			const rule = rules[Math.floor(chaosOptions.rng() * rules.length)];
			return rule(req, res, next);
		}
		return next();
	};
}

export { chaos, Rules };
